### PR TITLE
Remove the need of anyuid permission on OpenShift

### PR DESCRIPTION
**Pull request by jwendell**
- Remove the hard-coded usages of `runAsUser` and `runAsGroup` in charts
- Where necessary, replace them with a helm field `.ProxyUID` and `.ProxyGID`.
- When running in OpenShift, these fields will be set at runtime based on annotations present on the pod's namespace.
- When not on OpenShift (strictly speaking, when such annotations are not present), these values fallback to the current, defauls value of 1337.

With this change we can remove the extra steps related to `anyuid` in the [OpenShift setup page](https://istio.io/v1.17/docs/setup/platform-setup/openshift/).

### CODE REVIEW

File: pkg/kube/inject/openshift.go
Comment (jacob-delgado) [inline pkg/kube/inject/openshift.go:88]: similar to the log statement above you should print out `securityv1.UIDRangeAnnotation`

File: pkg/kube/inject/webhook.go
Comment (jacob-delgado) [inline pkg/kube/inject/webhook.go:996]: I think I tried something similar once and there was concern of api calls to apiserver. I'll leave this up to the networking maintainers.

**Review by jacob-delgado — COMMENTED** [review]
LGTM, but see my concern about adding the kubeclient to the webhook. I'll defer that to the networking maintainers. This would definitely help users of OpenShift so I'm in favor.

File: pkg/kube/inject/webhook.go
Comment (howardjohn) [reply pkg/kube/inject/webhook.go:996]: This is actually using an informer under the hood but we need to make one and use it everywhere rather than create one per request. Also need to check it's synced.

Can we disable this when not on open shift somehow?

File: pkg/kube/inject/openshift.go
Comment (jwendell) [reply pkg/kube/inject/openshift.go:88]: This is a copy of upstream (see the header in the file). I'd keep the differences minimum.

File: go.mod
Comment (howardjohn) [inline go.mod:66]: I would rather not bring in this large and dependency-heavy dependency just to bring in ~20 lines of code. Can we just copy them over?

File: manifests/charts/istio-control/istio-discovery/templates/deployment.yaml
Comment (howardjohn) [inline manifests/charts/istio-control/istio-discovery/templates/deployment.yaml:181]: Why is this removed? This will not be added back by injection, seems like a regression for non-openshift users

File: manifests/charts/istio-operator/templates/deployment.yaml
Comment (howardjohn) [inline manifests/charts/istio-operator/templates/deployment.yaml:43]: Why is this removed?

File: pilot/docker/Dockerfile.proxyv2
Comment (howardjohn) [inline pilot/docker/Dockerfile.proxyv2:44]: This is probably a breaking change

File: pkg/kube/inject/inject.go
Comment (howardjohn) [inline pkg/kube/inject/inject.go:868]: This explains what the code is doing but not _why_. Can you explain the openshift specific reasoning, in particular why "Max" is used?

File: pkg/kube/inject/webhook.go
Comment (howardjohn) [inline pkg/kube/inject/webhook.go:230]:
```
// addStartFunc appends a function to be run. These are run synchronously in order,
// so the function should start a go routine if it needs to do anything blocking
```

So we should not be blocking like this. Instead we would need run `wh.Run` as `go wh.Run` and add a HasSynced on webhook that polls and blocks readiness I think.

This is also blocking the wrong thing I think - this blocks the watcher, but the injection server is already started and marked ready at L86 of sidecarinjector.go. Technically I think this would work out since this will block the 'global readiness' but we should be cautious of the webhook server vs webhook patcher difference

**Comment by howardjohn** [comment]
Does openshift block the UIDs at Deployment level or only Pod level?

**Comment by jwendell** [comment]
> Does openshift block the UIDs at Deployment level or only Pod level?

Replicaset level.

File: pilot/docker/Dockerfile.proxyv2
Comment (jwendell) [reply pilot/docker/Dockerfile.proxyv2:44]: Why? All tests passed.

File: pkg/kube/client.go
Comment (jwendell) [reply pkg/kube/client.go:429]: IDK. This is what we are using in our fork. If you have any ideas, please let me know. I'll research more.

File: pkg/kube/inject/inject.go
Comment (jwendell) [reply pkg/kube/inject/inject.go:868]: Sorry, there's no specifics here. We could pick the first value, the last, or any random value. At first I'd thought using random, but since it doesn't really matters, I avoided calling a random function and just picked the max.

File: pilot/docker/Dockerfile.proxyv2
Comment (howardjohn) [reply pilot/docker/Dockerfile.proxyv2:44]: In general, because of the unknown unknowns.

Unfortunately we do not test every way everyone is using this... but I am certain there are people running the "Kubernetes YAML" from https://istio.io/latest/docs/setup/additional-setup/gateway/#deploying-a-gateway before we added the runAsUser section though. This change will switch them from root to 1337. Which, in general, is good - but the same people are also binding to port `80` which requires root,

File: pkg/kube/client.go
Comment (howardjohn) [reply pkg/kube/client.go:429]: For some other areas we detect GKE by the version (version is like `v1.24.14-gke.1200`). We already fetch the version so its 'free'.

If we do need to use discovery, is `ServerResourcesForGroupVersion` feasible to do a more specific lookup? That could be the difference between 1 API call vs 100s (https://blog.upbound.io/scaling-kubernetes-to-thousands-of-crds)

Comment (jwendell) [reply pkg/kube/client.go:429]: We can't use the version, as it's similar to vanilla k8s.

Another approach (cheapest one I guess) is to have a boolean variable eg (`OPENSHIFT`) defined in the helm charts, for the istiod deployment. It could be set using helm capabilities or freely using values - in which case could be used in the openshift profile). Is that a good option? Are we trying to get rid of such env variables?

Comment (howardjohn) [reply pkg/kube/client.go:429]: In general just trying to get rid of feature flags that are intended to be temporary but live forever. We have a variety of env var config, like `REVISION`, that will live forever. Let me think if there is a better way. In general `platform=openshift` may be more future proof (I can imagine needing a GCP specific thing in the future, etc)

Comment (jwendell) [reply pkg/kube/client.go]: I've switched to the new `platform.IsOpenShift()`

File: pkg/kube/inject/webhook.go
Comment (jwendell) [reply pkg/kube/inject/webhook.go]: I've added `HasSynced()` in the webhook and I'm calling it from `server .cachesSynced()`.

**Comment by howardjohn** [comment]
> > I am a bit worried these changes may have some subtle negative impact on non-openshift users. This is a bit vague since I don't have any concrete criticisms, more of the unknown-unknown of the template changes
> 
> The general idea is to remove hardcoded values in `RunAs` fields in the deployments. If this field is absent, Kubernetes will honor the `USER` value of the container image. OpenShift will automatically assign a value derived from an range defined in a namespace annotation.
> 
> If we change the `USER` field in the base proxy image to be `1337` - which we did in this PR - we can remove the `RunAs` declarations in some places. Kubernetes will use 1337 and OpenShift will assign a value at runtime.

The issue is if users are currently running the images without explicitly setting runas, they are now changed from `0` to `1337`. See https://github.com/istio/istio/pull/45394#discussion_r1227012907 - I am 100% certain this will break some of our users.

To be fair - its a good end result, we don't want users running as root. But I am a bit worried about the breakage.

Does openshift even need it? Can we just keep 1337 hardcoded as-is, and then override with the max UID range on openshift as added in this PR?

File: go.mod
Comment (jwendell) [reply go.mod:66]: done

File: manifests/charts/istio-control/istio-discovery/templates/deployment.yaml
Comment (jwendell) [reply manifests/charts/istio-control/istio-discovery/templates/deployment.yaml:187]: It looks like these files don't get parsed by the template, so we can't use `{{ .ProxyUID }}` here, we can only rely on `.Values`. Right?

File: manifests/charts/istio-operator/templates/deployment.yaml
Comment (jwendell) [reply manifests/charts/istio-operator/templates/deployment.yaml:43]: (same question as above)

File: manifests/charts/istio-control/istio-discovery/templates/deployment.yaml
Comment (howardjohn) [reply manifests/charts/istio-control/istio-discovery/templates/deployment.yaml:187]: Yes, they are not injected. They have nothing to do with ProxyUID technically either since they are not proxies... although maybe we want similar logic. 

AFAIK this is fixing openshift but in the process, making stock Kubernetes run as root which is not what we want.

Comment (jwendell) [reply manifests/charts/istio-control/istio-discovery/templates/deployment.yaml:187]: > AFAIK this is fixing openshift but in the process, making stock Kubernetes run as root which is not what we want.

No because the pilot image runs as user 1337.

Comment (jwendell) [reply manifests/charts/istio-control/istio-discovery/templates/deployment.yaml:187]: I mean, we are safe to remove these lines because https://github.com/istio/istio/blob/master/pilot/docker/Dockerfile.pilot#L25 and https://github.com/istio/istio/blob/master/operator/docker/Dockerfile.operator#L25

File: pilot/docker/Dockerfile.proxyv2
Comment (jwendell) [reply pilot/docker/Dockerfile.proxyv2:44]: Can't we add an upgrade note about this breaking change? Or do you consider this a so severe breakage that we should avoid at all costs?

File: manifests/charts/istio-control/istio-discovery/templates/deployment.yaml
Comment (howardjohn) [reply manifests/charts/istio-control/istio-discovery/templates/deployment.yaml:187]: ah right. Let me double check but I think you are correct

File: manifests/charts/gateways/istio-ingress/templates/deployment.yaml
Comment (howardjohn) [inline manifests/charts/gateways/istio-ingress/templates/deployment.yaml:56]: This breaks ephemeral containers: https://istio.io/latest/docs/ops/configuration/security/harden-docker-images/#debugging.

Before: container comes up as user 1337, works fine.
After: container comes up as user of the dockerfile, which could be `0` (it is for istio/base - others can be used), which is blocked by `runAsNonRoot: true`

File: pilot/docker/Dockerfile.proxyv2
Comment (howardjohn) [reply pilot/docker/Dockerfile.proxyv2:44]: This seems likely to cause outages when upgrades occur, especially since the gateway admin may not even be the one upgrading Istio - so they may not have even read the release note

File: manifests/charts/gateways/istio-ingress/templates/deployment.yaml
Comment (jwendell) [reply manifests/charts/gateways/istio-ingress/templates/deployment.yaml:56]: Fixed

File: manifests/charts/istio-control/istio-discovery/templates/deployment.yaml
Comment (jwendell) [reply manifests/charts/istio-control/istio-discovery/templates/deployment.yaml:192]: I think we can resolve this conversation?

File: manifests/charts/istio-operator/templates/deployment.yaml
Comment (jwendell) [reply manifests/charts/istio-operator/templates/deployment.yaml:43]: I think we can resolve this conversation?

**Comment by jwendell** [comment]
I reverted the `+ USER 1337:1337` change I had done in proxy dockerfile. Hopefully changes now are safe.

File: manifests/charts/gateways/istio-egress/templates/deployment.yaml
Comment (howardjohn) [inline manifests/charts/gateways/istio-egress/templates/deployment.yaml:56]: Its slightly awkward/inconsistent that we have an "openshift" profile and an "openshift" platform, and some values are controlled by one vs the other. 

But I think ok since otherwise we would need to add some more values here

**Review by howardjohn — APPROVED** [review]

**Comment by howardjohn** [comment]
@jwendell you have a do-not-merge, LGTM to merge when you are ready

**Comment by kfaseela** [comment]
@jwendell : I am trying to run integration tests locally on my mac, and starts getting the below error today. This used to work before. Do you think it is related?

```
2023-08-30T08:16:27.591923Z	error	Pod injection failed: failed to run injection template: template: inject:88:9: executing "inject" at <.ProxyUID>: can't evaluate field ProxyUID in type *inject.SidecarTemplateData
2023-08-30T08:16:27.996153Z	info	Sidecar injection request for echo1-5-51122/a-v1-7868db5759-***** (actual name not yet known)
2023-08-30T08:16:27.998146Z	error	Invalid template: template: inject:88:9: executing "inject" at <.ProxyUID>: can't evaluate field ProxyUID in type *inject.SidecarTemplateData
```

**Comment by kfaseela** [comment]
> @jwendell : I am trying to run integration tests locally on my mac, and starts getting the below error today. This used to work before. Do you think it is related?
> 
> ```
> 2023-08-30T08:16:27.591923Z	error	Pod injection failed: failed to run injection template: template: inject:88:9: executing "inject" at <.ProxyUID>: can't evaluate field ProxyUID in type *inject.SidecarTemplateData
> 2023-08-30T08:16:27.996153Z	info	Sidecar injection request for echo1-5-51122/a-v1-7868db5759-***** (actual name not yet known)
> 2023-08-30T08:16:27.998146Z	error	Invalid template: template: inject:88:9: executing "inject" at <.ProxyUID>: can't evaluate field ProxyUID in type *inject.SidecarTemplateData
> ```

ignore, things are working fine when I use latest test image.

**Pull request merged**
